[angjozee] iP

angjozee opened the pull request:
### **🏛️ Plato Task Manager**
“The beginning is the most important part of the work.” – Plato

Plato Task Manager is a simple, efficient, and text-based task manager designed to help you track your to-dos, deadlines, and events with ease.

**Features**
Task Management – Add, list, mark, and delete tasks
Find Command – Search for tasks by keyword
Persistent Storage – Saves tasks between sessions

**How to Get Started**
1. Download the latest release [here](https://github.com/nus-cs2103-AY2425S2/ip/pull/268#).
2. Run the program using:
`java -jar plato.jar`
3. Add your tasks and start managing them effectively!

**Code Example**
Plato Task Manager is built around the Plato class, which initializes and runs the application.

```
public class Plato {
    public static void main(String[] args) {
        new Plato("./data/tasks.txt").run();
    }
}
```

**Task List**
 Implement core task functionalities
 Add find command
 Improve error handling and UI

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename Duke to Plato and implement custom chatbot greeting

- Replaced the `Duke` chatbot with a new chatbot named `Plato`.
- Updated the greeting and farewell messages to align with Plato's unique personality.
- Removed the Duke ASCII logo and all references to Duke.
- Updated horizontal lines

Commit:
Add Echo functionality to Plato

- Implemented functionality to echo user commands
- Exit condition updated to recognise "Farewell" as the termination command

Commit:
Add task storage and listing functionality to Plato chatbot

- Implemented the ability to store user-entered tasks in memory
- Added functionality to display all stored tasks using the "tasks?" command.

Commit:
Add mark and unmark functionality to Plato chatbot

- Implemented 'mark <index>' command to mark tasks as done.
- Implemented 'unmark <index>' command to mark tasks as not done.
- Enhanced tasks? command to display tasks with status icons ([X] for done, [ ] for not done).
- Introduced 'Task' class to represent tasks with properties:
  - 'description': Task details.
  - 'isDone': Completion status.
  - Methods: 'markAsDone()', 'markAsNotDone()', 'getStatusIcon()'.

Commit:
Add Task Management Features with Event, Deadline, and ToDo Classes

- Implemented 'Event', 'Deadline', and 'Task' classes with inheritance.
- Enhanced user interaction with task-specific commands ('deadline', 'todo', 'event').

Commit:
Add input/output redirection and test cases for task management

- Implemented functionality to read user input from input.txt.
- Redirected program output to actual.txt for testing.
- Added test cases for ToDo, Deadline, and Event tasks.
- Prepared expected.txt to validate program output against expected results.
- Fixed typos and whitespace issues

Commit:
Add input validation and error handling for task commands

- Implemented validation for todo, deadline, and event commands to ensure non-empty descriptions and proper formatting.
- Added range checks for mark and unmark commands to handle invalid or out-of-range task numbers.
- Introduced user-friendly error messages for incorrect inputs using PlatoException.
- Added handling for unknown commands with descriptive error messages.
- Updated continue statements to ensure proper flow control after handling each command.

Commit:
Add delete command to remove tasks from the list

- Implemented delete command to remove tasks by index.
- Validated task indices to handle out-of-range and non-integer inputs.
- Updated task list dynamically using ArrayList<Task>.
- Printed details of the removed task and updated task count.

Commit:
Refactor code to use TaskType enum for task types

- Added a TaskType enum to represent task types (TODO, DEADLINE, EVENT).
- Updated the Task class to include a TaskType field for better structure and type safety.
- Refactored ToDo, Deadline, and Event classes to use TaskType.

Commit:
Add storage for tasks

- Ensure tasks are saved after every addition, modification, or deletion
- Handle file creation during loading

Commit:
Add date and time handling for deadlines and events

- Store deadlines and events using java.time.LocalDateTime for proper date/time parsing
- Accept user input in the format yyyy-MM-dd HHmm (e.g., 2019-12-02 1800)
- Format output dates as MMM dd yyyy, h:mm a (e.g., Dec 02 2019, 6:00 PM)
- Update file storage to save and load dates using proper formats

Commit:
Improve code structure with OOP design and command-based architecture

- Extract task-related logic into individual `Command` classes for maintainability
  - AddCommand: Handles adding ToDo, Deadline, and Event tasks
  - MarkCommand: Handles marking and unmarking tasks
  - DeleteCommand: Handles task deletion
  - ListCommand: Displays the list of tasks
  - ExitCommand: Exits the program
- Introduce an abstract `Command` class to encapsulate command execution logic
- Refactor core logic into `TaskList`, `Storage`, and `Ui` classes to improve modularity

Commit:
Organise project into packages for better modularity

- Create packages: command, exception, model, parser, storage, ui, and utils
- Move related classes into corresponding packages:
  - Command classes to command package (AddCommand, DeleteCommand, etc.)
  - Task-related classes to model package (Task, ToDo, Deadline, Event)
  - Parser and exception handling classes to parser and exception packages
  - Storage-related classes to storage package (Storage, PlatoFileWriter)
  - User interaction classes to ui package (Ui)
- Update package declarations and imports across all files

Commit:
Merge branch 'add-gradle-support'

Hackari commented on src/main/java/plato/command/AddCommand.java:
Static variables should be in SCREAMING_SNAKE_CASE! For the DateTimeFormatter at line 15.

Hackari commented on src/main/java/plato/model/ToDo.java:
I understand what toFileFormat() means, but it seems that we should be using verbs for our method names. For example, convertToFileFormat() might be better! (I made the same mistake by writing toData() 😞 )

Hackari reviewed (commented):
LGTM

githubcsuser commented on src/main/java/Plato.java:
```diff
@@ -8,15 +10,30 @@ public static void main(String[] args) {
+                }
+                System.out.println("______________________________________");
+            }
+            else {
```
Should "else" be inline with the closing brace i.e., "} else {" ?

githubcsuser commented on src/main/java/Plato.java:
```diff
@@ -8,15 +10,30 @@ public static void main(String[] args) {
                 break;
             }
+            if (userInput.equals("tasks?")) {
+                System.out.println("______________________________________");
```
Would it be helpful to store the separator "_______" line as a constant (e.g., String LINE = "___") to reduce repetition in println statements? What if it was written as ..println(LINE + "Farewell. Hope...!" + LINE);...

githubcsuser commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Should the field be private to enforce encapsulation?

githubcsuser commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+public class Deadline extends Task {
+
+    protected String by;
```
Would it be good to change these protected fields to private fields?

githubcsuser commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,36 @@
+______________________________________
+______________________________________
+Farewell. Hope to see you soon!
+______________________________________
```
Looks alright!

githubcsuser reviewed (commented):
Just some minor code changes to consider.

jonaturn commented on src/main/java/plato/Plato.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public static void main(String[] args) {
+        new Plato("./data/tasks.txt").run();
```
could this reading of tasks.txt be put into your IO handling class instead?

jonaturn commented on src/main/java/plato/model/TaskType.java:
```diff
@@ -0,0 +1,5 @@
+package plato.model;
```
I like the use of enum to decide class type

jonaturn commented on src/main/java/plato/storage/PlatoFileWriter.java:
```diff
@@ -0,0 +1,84 @@
+            if (!directory.exists()) {
+                directory.mkdir();
+            }
+            try (BufferedWriter bw = new BufferedWriter(new FileWriter("./data/plato.txt"))) {
```
I like how you use buffered write instead of system print lines

jonaturn commented on src/main/java/plato/storage/PlatoFileWriter.java:
```diff
@@ -0,0 +1,84 @@
+                    task = new ToDo(parts[2]);
+                    break;
+                case "D":
+                    String[] deadlineParts = parts[2].split(" \\| ");
```
Could a more descriptive name be used for your description string parts be used here?

jonaturn commented on src/main/java/plato/command/DeleteCommand.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws PlatoException {
```
I like the use of verbs in your method names, following the naming convention of this course

Commit:
Add JUnit tests for AddCommand and Command classes

- Added unit tests for AddCommand to validate ToDo, Deadline, and Event task creation.
- Included tests for valid and invalid date formats and proper exception handling.
- Added tests for Command class behaviour using mock implementations.

Commit:
Add Javadoc comments to Event class for improved documentation and clarity

Commit:
Add JavaDoc comments for ListCommand.java

Commit:
Add Javadoc comments for Parser and FindCommand classes to improve documentation and clarity

Commit:
Fix packaging issues

Commit:
Merge branch 'branch-A-Jar'

Commit:
Apply Checkstyle fixes and enforce coding standards

Commit:
Fix event task parsing and file format

Commit:
Add FindCommand to search for tasks by keyword

Commit:
Refactor Command execution to return responses & enable JavaFX exit on "bye"

Commit:
Plato, MainWindow: Add assertions to enforce assumptions

Some critical assumptions in the code were not explicitly checked,
which could lead to unexpected runtime errors. For instance, user input
and UI components were assumed to be initialized but were never
validated.

This lack of validation could lead to `NullPointerException` or
unexpected behavior if components were not properly set up.

Let's add assertions in key areas to enforce these assumptions and
improve code robustness:

- Ensure `filePath` in `Plato` constructor is non-null and non-empty
- Validate that `tasks` and `command` are properly initialized
- Assert `userInput`, `dialogContainer`, `scrollPane`, and `sendButton`
  are initialized in `MainWindow`
- Ensure that `getResponse` input and return values are not null

These assertions will help catch logical errors early in development
without affecting production performance, as assertions can be disabled
in runtime.

Commit:
Add assertions

Commit:
Improve code quality

Commit:
Merge branch 'branch-A-Assertions'

Commit:
Fix priority parsing, storage, and display in tasks

Previously, task priorities were not correctly identified when adding a
task. If a priority (e.g., HIGH, MEDIUM, LOW) was provided, it was being
treated as part of the task description. Additionally, the priority
was stored incorrectly in tasks.txt for Deadline and Event tasks,
leading to duplicate priority values in both file storage and display.

This commit implements the following fixes:

- **Correctly parses priority levels** when adding tasks:
  - Users can now specify priority at the end of their command.
  - If no priority is provided, the task defaults to MEDIUM.
  - Example: `todo Read book HIGH` correctly assigns HIGH priority.

- **Fixes duplicate priority in tasks.txt**:
  - `Deadline` and `Event` tasks previously appended an extra priority.
  - Now, tasks are saved correctly:
    - Before: `D ||   || sleep || MEDIUM || 2003-02-09 0012 || MEDIUM`
    - After:  `D ||   || sleep || 2003-02-09 0012 || MEDIUM`

- **Fixes duplicate priority in console output**:
  - `toString()` for `Deadline` and `Event` was printing priority twice.
  - Now, priority appears only once when displaying tasks.

- **Refactored `AddCommand` to properly extract priority**:
  - A helper method `isValidPriority()` ensures only valid priorities are parsed.
  - Prevents cases where an invalid priority is mistaken for part of the description.

Commit:
Improve UI styling and update FXML layout

Commit:
Update README
